File: frontend/src/app/services/merchandise.service.ts
import { Injectable } from '@angular/core';
import {Observable} from 'rxjs';
import {Merchandise} from '../dtos/merchandise';
import {HttpClient, HttpParams} from '@angular/common/http';
import {Globals} from '../global/globals';

@Injectable({
  providedIn: 'root'
})
export class MerchandiseService {

  private merchandiseBaseUri: string = this.globals.backendUri + '/merchandise';

  constructor(private httpClient: HttpClient, private globals: Globals) {
  }

  /**
   * Get all merchandise items that are available as bonuses for a specified amount of bonus points.
   *
   * @param bonusPoints specifies the upper limit for the bonus points.
   * @param page specifies the page number index
   */
  getAvailableBonus(bonusPoints: number, page: number): Observable<Merchandise[]> {
    const bonus = bonusPoints.toString();
    console.log('Get available merchandise for: ' + bonus + ' bonus points');
    const params = new HttpParams().set('bonus', bonus);
    return this.httpClient.get<Merchandise[]>(this.merchandiseBaseUri + '/available' + '/pages/' + page , {params});
  }

  /**
   * Get all merchandise items that belong to the specified event.
   *
   * @param eventId specifies the event.
   */
  getMerchandiseForEvent(eventId: number): Observable<Merchandise[]> {
    console.log('Get merchandise for event with id: ' + eventId);
    return this.httpClient.get<Merchandise[]>(this.merchandiseBaseUri + '/event/' + eventId);
  }

  /**
   * Get all merchandise items.
   */
  getAllMerchandise(page: number): Observable<Merchandise[]> {
    console.log('Get all merchandise items');
    return this.httpClient.get<Merchandise[]>(this.merchandiseBaseUri + '/pages/' + page);
  }

  /**
   * Get one specified merchandise item.
   * @param id specifies the merchandise item.
   */
  getMerchandiseItemById(id: number): Observable<Merchandise> {
    console.log('Get merchandise item with id: ' + id);
    return this.httpClient.get<Merchandise>(this.merchandiseBaseUri + '/' + id);
  }

  /**
   * Persists merchandise in the backend
   * @param merchandise to persist
   */
  createMerchandise(merchandise: Merchandise): Observable<Merchandise> {
    console.log('Create merchandise item with the title: ' + merchandise.title);
    return this.httpClient.post<Merchandise>(this.merchandiseBaseUri, merchandise);
  }

  /**
   * save image to a specified merchandise item
   * @param id specifies the merchandise item
   * @param formData image
   */
  public uploadImage(id: number, formData) {
    return this.httpClient.post<any>(this.merchandiseBaseUri + '/' + id + '/upload', formData);
  }

  /**
   * Loads index from the backend where the images should be saved
   */
  getIndexForImage(): Observable<number> {
    return this.httpClient.get<number>(this.merchandiseBaseUri + '/index');
  }

  getImages(id: number) {
    console.log('Load image for merchandise item with id: ' + id);
    return this.httpClient.get(this.merchandiseBaseUri + '/' + id + '/image', {responseType: 'text'});
  }
}

File: backend/src/main/java/at/ac/tuwien/sepm/groupphase/backend/service/MerchandiseService.java
package at.ac.tuwien.sepm.groupphase.backend.service;

import at.ac.tuwien.sepm.groupphase.backend.entity.Merchandise;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.web.multipart.MultipartFile;

import java.util.List;

public interface MerchandiseService {

    /**
     * Save a merchandise item to the repository.
     *
     * @param merchandise the merchandise item to be saved, containing all the necessary information
     * @return the merchandise item that has been saved to the repository
     */
    Merchandise save(Merchandise merchandise);

    /**
     * Get all merchandise items.
     *
     * @param pageable page configuration
     * @return List of all the stored merchandise items
     */
    Page<Merchandise> getAllMerchandiseItems(Pageable pageable);

    /**
     * Get all the merchandise items for a specified event.
     *
     * @param eventId specifies the event
     * @return List of merchandise items that belong to the specified event
     */
    List<Merchandise> getMerchandiseForEvent(Long eventId);

    /**
     * Get all merchandise items that are available for certain bonus points.
     *
     * @param bonusPoints the upper limit for the bonus points
     * @param pageable page configuration
     * @return List of merchandise items which bonus points are <= the specified bonus points
     */
    Page<Merchandise> getAvailableMerchandise(Long bonusPoints, Pageable pageable);

    /**
     * Get specified merchandise item.
     *
     * @param id specifies the merchandise item
     * @return the merchandise item corresponding to the specified id
     */
    Merchandise getMerchandiseById(Long id);

    /**
     * Get index for the next merchandise item to be added.
     * @return index for the next merchandise item
     */
    Long getNextIndex();

    /**
     * Save an image to a specified merchandise item to the repository.
     * @param id specifies the merchandise item to which the image is stored
     * @param file the image to be stored
     */
    void uploadImage(Long id, MultipartFile file);

    /**
     * Get the image for a specified merchandise item.
     * @param id specifies the merchandise item
     * @return base64 encoded image
     */
    String getImage(Long id);
}
